### PR TITLE
[GOR-50] Location End Point

### PR DESCRIPTION
**Pull request by kevo1ution**

### CODE REVIEW

**Comment by kevo1ution** [comment]
@sarahannali I decided not to include the virtual values because the ".lean()" method saves a lot of time and memory (around 10x), and I think the benefits are really good. This means on the app we will need to get the vegan rating by dividing total and count

File: src/api/routes/location.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+const router = express.Router();
+
+router.get('/',
+  celebrate({
+    [Segments.QUERY]: Joi.object().keys({
+      filter: Joi.object().keys({
+        lat: Joi.number().max(90).min(-90).required(),
+        long: Joi.number().max(180).min(-180).required(),
+        radius: Joi.number().max(MAX_DISTANCE_METERS).min(0).required(),
+        veganRating: Joi.number().valid(2.5, 3.5, 4.5),
```
Comment (sarahannali) [inline src/api/routes/location.js]: i thought we were making this 3.5, 4.0, and 4.5

File: src/api/routes/location.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        open: Joi.number().integer().max(7 * 24 * 60).min(0),
```
Comment (sarahannali) [inline src/api/routes/location.js:20]: what does this represent?

File: src/api/routes/location.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        $maxDistance: filter.radius,
+        $minDistance: 0,
+      },
+    };
+
+    // Optional filters
+    if (filter.open) {
+      mongoFilter.openingHours = {
+        $elemMatch: {
+          start: { $lte: filter.open },
```
Comment (sarahannali) [inline src/api/routes/location.js:52]: also this part, why is start less than open and end greater than open

File: test/util/mongoose.js [hunk 4]
```diff
@@ -12,6 +12,7 @@ class TestDB {
     await Promise.all(modelNames.map(async (modelName) => {
       if (seedData[modelName]) {
         const Model = Models[modelName];
+        await Model.ensureIndexes();
```
Comment (sarahannali) [inline test/util/mongoose.js:15]: [here](https://mongoosejs.com/docs/deprecations.html) it talks about ensureIndexes deprecation stuff does this matter for us do we need to set useCreateIndex to true. also i don't really understand what this does

File: src/api/routes/location.js (hunk 2 above)
Comment (kevo1ution) [reply src/api/routes/location.js:20]: open is going to be the users current time that they send. So for example if it is 6:00 am on a Sunday then the front end sends open = 6 * 60 = 360 to get all locations that are currently open at that time (still within the lat long and radius specified)

File: src/api/routes/location.js (hunk 3 above)
Comment (kevo1ution) [reply src/api/routes/location.js:52]: we are searching to make sure the open (aka current user time) is in between the start and end of an interval. If there exists an array that satisfies this then that means the restaurant is open at that time.

File: test/util/mongoose.js (hunk 4 above)
Comment (kevo1ution) [reply test/util/mongoose.js:15]: https://github.com/nodkz/mongodb-memory-server/issues/102
I had issues with the test cases not having created the 2dsphere index on the lat and long properties of Location. The ensureIndexes forces mongoose to create those indexes for each model. I'm not sure if there is any other work around. It is also for test cases so maybe deprecation is not as serious of an issue

**Review by sarahannali — APPROVED** [review]

**Pull request merged**